datalad/local/gitcredential_datalad.py

```diff
@@ -0,0 +1,181 @@
+import sys
+
+from datalad.downloaders import UserPassword
+from datalad.downloaders.providers import (
+    AUTHENTICATION_TYPES,
+    Providers,
+)
+from datalad.local.gitcredential import _credspec2dict
+from datalad.utils import Path
+
+git_credential_datalad_help = """\
+Git credential interface to DataLad's credential management system.
+
+Only DataLad's UserPassword-type credentials are supported. This helper
+passes standard 'get', 'store', and 'erase' actions on to the respective
+interfaces in DataLad.
+
+usage: git credential-datalad [option] <action>
+
+options:
+    -h                   Show this help.
+    --non-interactive    Don't ask for user confirmation when storing to
+                         DataLad's credential system. This may fail if default
+                         names result in a conflict with existing ones.
+                         This mode is used for DataLad's CI tests.
+"""
+
+
+def git_credential_datalad():
+    """Entrypoint to query DataLad's credentials via git-credential
+    """
+
+    if not 2 <= len(sys.argv) <= 3 \
+            or sys.argv[1] == "-h" \
+            or sys.argv[1] != "--non-interactive" \
+            or sys.argv[-1] not in ['get', 'store', 'erase']:
+        help_explicit = sys.argv[1] == "-h"
+        print(git_credential_datalad_help,
+              file=sys.stdout if help_explicit else sys.stderr)
+        sys.exit(0 if help_explicit else 1)
+
+    interactive = False if sys.argv[1] == "--non-interactive" else True
+    action = sys.argv[-1]
+    attrs = _credspec2dict(sys.stdin)
+
+    # This helper is intended to be called by git. While git-credential takes a
+    # `url` property of the description, what it passes on is `protocol`, `host`
+    # and potentially `path` (if instructed to useHttpPath by config).
+    # For communication with datalad's credential system, we need to reconstruct
+    # the url, though.
+    assert 'protocol' in attrs.keys()
+    assert 'host' in attrs.keys()
+    if 'url' not in attrs:
+        attrs['url'] = "{protocol}://{host}{path}".format(
+            protocol=attrs.get('protocol'),
+            host=attrs.get('host'),
+            path="/{}".format(attrs.get('path')) if 'path' in attrs else ""
+        )
+
+    # Get datalad's provider configs.
+    providers = Providers.from_config_files()
+
+    if action == 'get':
```

Comment:
What do you think about de-spaghettifying this monolithic helper and put its actions into dedicated helper functions that could have a dedicated unittest without having to go through the complexity of the entire system for testing?

Reply:
Yes, I agree. Will do.